### PR TITLE
Add boost and claim endpoints

### PR DESCRIPTION
**Pull request by taycaldwell**
Add minor update to claimed_boost schema to store contract address associated with each claim for NFT claims.

### CODE REVIEW

File: src/app/api/boosts/claim/route.ts [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+import { NextResponse, type NextRequest } from 'next/server';
+import '@/utils/helper';
+import { Network, Alchemy } from "alchemy-sdk";
+
+const settings = {
+    apiKey: process.env.ALCHEMY_ID,
+    network: Network.BASE_MAINNET,
+};
+
+const alchemy = new Alchemy(settings);
```
Comment (siwan-cb) [inline src/app/api/boosts/claim/route.ts:11]: doesn't this mean alchemy will be initialized every time we call /claim endpoint?
I think this should be initialized in Providers.tsx?

File: src/app/api/boosts/claim/route.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+async function ownsNFT(userAddress: string, contract: string) {
+    const response = await alchemy.nft.verifyNftOwnership(userAddress, contract);
+    return response;
+}
+
+async function hasToken(userAddress: string, contract: string) {
+    const contracts = [contract];
```
Comment (siwan-cb) [inline src/app/api/boosts/claim/route.ts:19]: not needed.

File: src/app/api/boosts/claim/route.ts [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+
+    const tokenBalanceRes = await alchemy.core.getTokenBalances(userAddress, contracts);
```
Comment (siwan-cb) [inline src/app/api/boosts/claim/route.ts:21]:
```suggestion
    const tokenBalanceRes = await alchemy.core.getTokenBalances(userAddress, [contract]);
```

File: src/app/api/boosts/claim/route.ts [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    const filterByBalance = tokenBalanceRes.tokenBalances.filter((balance: any) => Number(balance.tokenBalance) > 0);
```
Comment (siwan-cb) [inline src/app/api/boosts/claim/route.ts:22]: no any

Comment (siwan-cb) [reply src/app/api/boosts/claim/route.ts:22]: if tokenBalance is a bigint and if you try to convert to a number, you have a risk of overflow. Couldn't you just do `balance.tokenBalance > 0n`?

File: src/app/api/boosts/claim/route.ts [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    return filterByBalance.length > 0;
+}
+
+const prisma = new PrismaClient();
```
Comment (siwan-cb) [inline src/app/api/boosts/claim/route.ts:26]: ah we are also doing this everywhere in our API. This should really be a singleton.

cc @amit-shah-cb

File: src/app/api/boosts/claim/route.ts [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+  switch (boost.boost_type) {
+    case 'NFT':
+        verified = await ownsNFT(userAddress, contractAddress);
+        break;
+    case 'NFT_PER_MINT':
+        verified = await ownsNFT(userAddress, contractAddress);
+        break;
+    case 'TOKEN':
+        verified = await hasToken(userAddress, contractAddress);
+        break;
```
Comment (siwan-cb) [inline src/app/api/boosts/claim/route.ts:90]:
```suggestion
    case 'NFT':
    case 'NFT_PER_MINT':
    case 'TOKEN':
        verified = await ownsNFT(userAddress, contractAddress);
        break;
```

File: src/app/api/boosts/claim/route.ts [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    case 'RECURRING':
+        const activeClaims = boost.claimed_boost.filter(c => {
+            const refreshHours = boost.refresh_time ? boost.refresh_time.getHours() : 24;
+            const expirationTime = new Date(c.updated_at.getTime() + refreshHours * 60 * 60 * 1000);
```
Comment (siwan-cb) [inline src/app/api/boosts/claim/route.ts:94]: 60 * 60 * 1000 should be defined outside of this function as class variable

File: src/app/api/boosts/claim/route.ts [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+            return new Date() < expirationTime;
```
Comment (siwan-cb) [inline src/app/api/boosts/claim/route.ts:95]: does date object can do logical operation? shouldn't we use date-fns?

File: src/app/api/boosts/claim/route.ts [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+    const claimedBoost = await prisma.claimed_boost.create({
+        data: {
+          user_address: userAddress,
+          boost_id: boost.id,
+          game_id: gameId,
+          contract_address: contractAddress ? contractAddress : null
+        }
+      });
+      return NextResponse.json(claimedBoost);
+  } else {
```
Comment (siwan-cb) [inline src/app/api/boosts/claim/route.ts:111]: /nit else is not needed here.

File: src/app/api/boosts/claim/route.ts [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+    return new Response(`Unable to claim boost for boostId: ${boostId}, gameId: ${gameId}`, {
```
Comment (siwan-cb) [inline src/app/api/boosts/claim/route.ts:112]: I would actually check for `!verified` first.

File: src/app/api/boosts/route.ts [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+    switch (boost.boost_type) {
+        case 'NFT':
+            isClaimed = boost.contract_addresses.every(a => claimedAddresses.includes(a));
+            break;
+        case 'NFT_PER_MINT':
+            isClaimed = boost.contract_addresses.every(a => claimedAddresses.includes(a));
+            break;
+        case 'TOKEN':
+            isClaimed = boost.claimed_boost.every(c => c.boost_id === boost.id);
+            break;
```
Comment (siwan-cb) [inline src/app/api/boosts/route.ts:52]:
```suggestion
        case 'NFT':
        case 'NFT_PER_MINT':
        case 'TOKEN':
            isClaimed = boost.contract_addresses.every(a => claimedAddresses.includes(a));
            break;
```

File: src/app/api/boosts/route.ts [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+        case 'RECURRING':
+            const activeClaims = boost.claimed_boost.filter(c => {
+                const refreshHours = boost.refresh_time ? boost.refresh_time.getHours() : 24;
+                const expirationTime = new Date(c.updated_at.getTime() + refreshHours * 60 * 60 * 1000);
```
Comment (siwan-cb) [inline src/app/api/boosts/route.ts:56]: same regarding 60 * 60 * 1000

**Comment by siwan-cb** [comment]
on your next PR, could you rebase? this PR has commits from other people as well.

File: src/app/api/boosts/claim/route.ts [hunk 13]
```diff
@@ -0,0 +1,108 @@
+import { PrismaClient } from '@prisma/client';
```
Comment (amit-shah-cb) [inline src/app/api/boosts/claim/route.ts:1]: we are no longer using prisma as it is a bottleneck, please take a look at hoq to query or insert rows with supabase sdk. 

Currently supbase sdk is not enforcing strong type, you can make db changes and insert; we will migrate to use strong types now that scalability is potentially resolved.

Take a look at: /api/webhook/treasure-box/route.ts for example of initilaizing supabase sdk and inserting

File: src/app/api/boosts/claim/route.ts [hunk 14]
```diff
@@ -0,0 +1,108 @@
...
+  if (!userAddress || !gameId || !boostId) {
+    return new Response(
+      `Missing parameters: userAddress: ${userAddress}, gameId: ${gameId}, boostId: ${boostId}`,
+      {
+        status: 400,
+      }
+    );
+  }
+
+  const boost = await prisma.boost_configuration.findFirst({
```
Comment (amit-shah-cb) [inline src/app/api/boosts/claim/route.ts:43]: can be replaced with supabase.from("boost_configuration").select ....

File: src/app/api/boosts/route.ts [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+  if (!userAddress || !gameId) {
+    return new Response(
+      `Missing parameters: userAddress: ${userAddress}, gameId: ${gameId}`,
+      {
+        status: 400,
+      }
+    );
+  }
+
+  const boosts = await prisma.boost_configuration.findMany({
```
Comment (amit-shah-cb) [inline src/app/api/boosts/route.ts:22]: replace with supabase query

**Pull request closed without merging**